lib/src/public_interface.dart

```diff
@@ -0,0 +1,71 @@
+import 'package:yaml/yaml.dart';
+
+import './source_edit.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments and
+/// whitespaces.
+///
+/// YAML parsing is supported by `package:yaml`, and modifications are performed as
+/// string operations. Each time a modification takes place via one of the public
+/// methods, we calculate the expected final result, and parse the result YAML string,
+/// and ensure the two YAML trees match. Users may define the default settings to be
+/// applied to these string modifications. Note however that these settings only apply
+/// to portions of the YAML that are modified by this class.
+///
+/// [1]: https://yaml.org/
+abstract class YamlEditBuilder {
+  /// The configuration to be used for the various string manipulation operations.
+  final YamlEditBuilderConfig config;
+
+  /// Returns a list of [SourceEdit]s detailing the string modifications that have
+  /// been made. Intended to be compatible with `package:analysis_server`.
+  List<SourceEdit> get edits;
+
+  YamlEditBuilder(String yaml, {this.config = const YamlEditBuilderConfig()});
+
+  /// Returns the current YAML string after the various modifications.
+  @override
+  String toString();
+
+  /// Returns the [YamlNode] present at the path.
+  YamlNode parseValueAt(Iterable<Object> path);
+
+  /// Sets [value] in the [path]. If the [path] is not accessible (e.g. it currently
+  /// does not exist in the document), an error will be thrown.
+  void setIn(Iterable<Object> path, Object value);
```

Comment:
We don't have to fix this here and now, but we should work to come up with a consistent naming convention for the mutation methods.

`In` signaling that the mutation happens in some _path_ might make sense.. I was tempted to use `At`.

Let's perhaps identify the core methods, the ones that aren't convenience methods.

 * For objects we have:
   (1) set key (add a new key, or overwrite an existing key)
   (2) remove key
 * For lists we have:
   (3) insert entry in list
   (4) set entry in list (overwriting an existing entry)
   (5) remove entry in list

Technically, speaking I suppose there is no difference between first removing an entry (5) and then adding an entry to a list (3).

-------

**Option A** we could do:
 * `void YamlEditBuilder.insert(Iterable<Object> path, Object value)`, where,
    * if `path` points to a key in an object we overwrite/add key/value pair,
    * if `path` points to an element in a list, we insert `value` in the list such that it has the position `path`.
      (hence, `YamlEditBuilder('{a: [1,2,3]}').insert([a, 0], 42)` will create `{a: [42, 1,2,3]}`, notice that the `value` inserted got index `0` in the list).
 * `void YamlEditBuilder.remove(Iterable<Object> path, Object value)`
    * If `path` points to a key in an object, we remove said key/value from the object.
    * If `path` points to an element in a list, we remove said element of the list.

With this API we could do all operations, correct?

Note. I'm not sure we should do this. Because `YamlEditBuilder('{a: [1,2,3]}').insert([a, 0], 42)` feels unintuitive.

--------

**Option B** we could do:
 * `void YamlEditBuilder.set(Iterable<Object> path, Object value)`,
    * if `path` points to a key in an object we overwrite/add key/value pair,
    * if `path` points to an element in a list we overwrite the element in the list.
 * `void YamlEditBuilder.insert(Iterable<Object> path, int index, Object value)`,
    * only works if `path` points to a list, then it inserts `value` at `index` in the list.
 * `void YamlEditBuilder.remove(Iterable<Object> path, Object value)`
    * If `path` points to a key in an object, we remove said key/value from the object.
    * If `path` points to an element in a list, we remove said element of the list.

The API is perhaps less concise, but more intuitive?

--------
In both cases we could consider postfixing API methods with `At`, `In` or `AtPath`.. Though given that almost all methods will behave the same maybe we don't need to?

In both cases we could also have convenience methods like `prepend`, `append`, more?

And are there other API designs we should consider?

note. it might be useful to indicate whether a method is purely convenience, or if it does something that can't be achieved by other means.

Reply:
So in coming up with the interface I preferred the second one, because it appeared more intuitive and more consistent with dart methods (if you think of set as applying `[]=`). Agree that almost all methods will behave the same, and we might not need a postfix - the motivation for that was more to remind people that a `path` is required, but the postfix does nothing much otherwise.

Could I clarify what you meant by this line:
> Technically, speaking I suppose there is no difference between first removing an entry (5) and then adding an entry to a list (3).

There was one other design I was contemplating, where instead of 
```dart
void YamlEditBuilder.set(Iterable<Object> path, Object value)

we had

void YamlEditBuilder.setInMap(Iterable<Object> mapPath, Object key, Object value)
void YamlEditBuilder.setInList(Iterable<Object> listPath, int index, Object value)
```

which is more verbose but allows us to be clearer with regards to what the collection at the end of the respective paths should be, and allows type-checking for index. Eventually decided on what I have just because I wasn't sure how much value this alternative provides.